crates/block-producer/src/tx_pool/mod.rs

```diff
@@ -0,0 +1,514 @@
+
+    batch_pool: BTreeMap<BatchId, InflightBatch>,
+
+    /// Batches which are ready to be included in a block.
+    ///
+    /// aka batches who's parent batches are already included in blocks.
+    batch_roots: BTreeSet<BatchId>,
+
+    /// Blocks which are inflight or completed but not yet considered stale.
+    block_pool: BTreeMap<BlockNumber, Vec<BatchId>>,
+
+    /// The next block we expect to complete.
+    next_completed_block: BlockNumber,
+}
+
+impl TransactionPool {
+    /// Complete barring todos.
+    pub fn add_transaction(
+        mut self,
+        transaction: ProvenTransaction,
+        mut inputs: TransactionInputs,
+    ) -> Result<(), AddTransactionError> {
+        // Ensure inputs aren't stale.
+        if inputs.current_block_height <= self.stale_block.0 {
+            return Err(AddTransactionError::StaleInputs {
+                input_block: BlockNumber(inputs.current_block_height),
+                stale_limit: self.stale_block,
+            });
+        }
+
+        let account_update = transaction.account_update();
+
+        // Inflight transactions upon which this new transaction depends due to building on their
+        // outputs.
+        let mut parents = BTreeSet::new();
+
+        // Merge inflight state with inputs.
+        //
+        // This gives us the latest applicable state for this transaction.
+        // TODO: notes and nullifiers.
+        if let Some((state, parent)) = self.account_state.get(&account_update.account_id()) {
+            parents.insert(*parent);
+            inputs.account_hash = Some(*state);
+        }
+
+        // Verify transaction input state.
+        // TODO: update notes and nullifiers.
+        if inputs.account_hash.unwrap_or_default() != account_update.init_state_hash() {
+            return Err(AddTransactionError::InvalidAccountState {
+                current: inputs.account_hash.unwrap_or_default(),
+                expected: account_update.init_state_hash(),
+            });
+        }
+
+        // Transaction is valid, update inflight state.
+        // TODO: update notes and nullifiers.
+        let tx_id = transaction.id();
+        self.account_state
+            .insert(transaction.account_id(), (account_update.final_state_hash(), tx_id));
+
+        // Inform parent's of their new child.
+        for parent in &parents {
+            self.tx_pool.get_mut(parent).expect("Parent must be in pool").add_child(tx_id);
+        }
+
+        // Insert transaction into pool and possibly as a root transaction.
+        self.tx_pool.insert(tx_id, InflightTransaction::new(transaction, parents));
+        self.try_root_transaction(tx_id);
+
+        Ok(())
+    }
+
+    /// Returns at most `count` transactions and a batch ID.
+    ///
+    /// Transactions are returned in a valid execution ordering.
+    ///
+    /// Returns `None` if no transactions are available.
+    pub fn select_batch(&mut self, count: usize) -> Option<(BatchId, Vec<Arc<ProvenTransaction>>)> {
```

Comment:
Why do we want to provide `count` as a parameter? I would have thought that it would be up to the logic in this method to figure out how many transactions to return (and that would be based on some config parameters of the transaction pool).

Also, maybe for the return type we should have a dedicated struct? Something like `SelectedBatch` or `ProposedBatch`.

Reply:
I was unsure where `count` should live; could be configured as a part of the pool or as a value in the batch producer. Logistically it made sense to me that it might be a constraint in the batch producer. But one could also argue it might form part of the transaction selection strategy.